sdk/python/lib/pulumi/provider/server.py

```diff
@@ -0,0 +1,189 @@
+"""Define gRPC plumbing to expose a custom user-defined `Provider`
+instance as a gRPC server so that it can be used as a Pulumi plugin.
+
+"""
+
+from typing import Dict, List, Optional
+import argparse
+import asyncio
+import sys
+
+import grpc
+import grpc.aio
+import google.protobuf.struct_pb2 as struct_pb2
+
+from pulumi.provider.provider import Provider, ConstructResult
+from pulumi.runtime import proto, rpc
+from pulumi.runtime.proto import provider_pb2_grpc, ResourceProviderServicer
+from pulumi.runtime.proto.provider_pb2 import ConstructRequest
+import pulumi
+import pulumi.resource
+import pulumi.runtime.config
+import pulumi.runtime.settings
+
+
+# _MAX_RPC_MESSAGE_SIZE raises the gRPC Max Message size from `4194304` (4mb) to `419430400` (400mb)
+_MAX_RPC_MESSAGE_SIZE = 1024 * 1024 * 400
+_GRPC_CHANNEL_OPTIONS = [('grpc.max_receive_message_length', _MAX_RPC_MESSAGE_SIZE)]
+_GRPC_WORKERS = 4
+
+
+class ProviderServicer(ResourceProviderServicer):
+    """Implements a subset of `ResourceProvider` methods to support
+    `Construct` and other methods invoked by the engine when the user
+    program creates a remote `ComponentResource` (with `remote=true`
+    in the constructor).
+
+    See `ResourceProvider` defined in `provider.proto`.
+
+    """
+
+    engine_address: str
+    provider: Provider
+    args: List[str]
+
+    async def Construct(self, request: ConstructRequest, context):  # pylint: disable=invalid-overridden-method
+        assert isinstance(request, ConstructRequest), \
+            f'request is not ConstructRequest but is {type(request)} instead'
+
+        # Should we take a lock so Construct calls are never executing
```

Comment:
Unresolved Q1

Reply:
We did this in Node.js as a short-term "solution" since the `Construct` call modifies global state. The downside is that it'll result in deadlock if you make a nested `construct` call within the same provider.

I believe the Python SDK has the same problem as Node.js (modifying global state in `Construct`).

We’re going to need to find a better long-term solution. For Node.js, we’ll likely end up making use of [`vm.createContext`](https://nodejs.org/api/vm.html#vm_vm_createcontext_contextobject_options) to run the user's `construct` in its own context (with it's own globals). I'm planning to try this out this week.

Go doesn't have this problem as each `Construct` call will get its own [`pulumi.Context`](https://github.com/pulumi/pulumi/blob/ee2f65510bd8129adffb18d7e198bf12c1b84fee/sdk/go/pulumi/context.go#L44).

The short-term thing for Python would be to do the same thing we did with Node.js (use locks to serialize `Construct` calls). That would at least ensure correctness when there are multiple concurrent `Construct` calls happening within the same process.

But I'm not sure what the right long-term solution will be for Python. Launch a new `python` process for each `Construct` call? Is there an equivalent of `vm.createContext` for Python?

Reply:
Ack on the short-term will implement that straight away. I can do some research on what we can do in Python for this long term and get back to you. At a first glance there is `threading.local()` so multithreaded programs can have per-thread state, which is a bit less overhead than a process; there's gotchas with Python threading though. I can see what async Python offers.

Would a "clean" solution of refactoring the SDK to not have to set the globals but pass parameters explicitly be too prohibitive?

Reply:
Maybe worth noting here is that currently, each RPC in python [runs in its own thread](https://github.com/pulumi/pulumi/blob/master/sdk/python/lib/pulumi/runtime/resource.py#L304) as we have not yet adopted grpc-python's [async API](https://grpc.github.io/grpc/python/grpc_asyncio.html). This has led to some [odd behavior](https://github.com/pulumi/pulumi/issues/6561) in the interaction of asyncio event loops and threads.

I'm not familiar enough with the code here to know if it's relevant, but just wanted to flag it in case it is.

Reply:
Follow-up - We'll need to [bump the version](https://github.com/grpc/grpc/releases/tag/v1.33.2#tag-select-menu-0e7b813a-9bc3-11eb-8090-b5026ca30356:~:text=Allow%20asyncio%20API%20to%20be%20imported%20as%20grpc.aio.) of grpc to use the async API in our [Pipfile](https://github.com/pulumi/pulumi/blob/master/sdk/python/Pipfile#L9)

Reply:
I'm adopting gRPC async API in this PR. Are there problems with it?

Reply:
> I'm adopting gRPC async API in this PR. Are there problems with it?

Not necessarily! I'm of the opinion that we should be adopting it for all of our usage of grpc across the codebase, so I'm glad to see it here. I just don't know if/how this will interact with the existing code in the rest of the python SDK. This isn't a "don't do it" but more like "I'm curious to see what happens".

Reply:
This new code would run inside a subprocess dedicated to wrapping user code that implements the new provider interface. I'm optimistic this will therefore not impact any existing code, but we should be comfortable maintaining it in the future.

Reply:
Yeah, I've been playing around with the async API in a local branch and would certainly like to adopt it elsewhere due to the aforementioned oddities with how our async code interacts with the RPCs. Thanks for taking a stab at it!

Reply:
Regarding Pipfile, as written it resolves to 1.37.0 so I think we're good? Do we have Pipfile.lock anywhere?

Reply:
> Regarding Pipfile, as written it resolves to 1.37.0

Right, but it is incorrect because you actually need `>=1.33.2`. For folks who have an existing version on their machines with a `Pipfile.lock` on a version prior to 1.33.2 wouldn't know to update. 

> Do we have Pipfile.lock anywhere?

We don't put the `Pipfile.lock` in source control because we target multiple versions of python and multiple platforms and the `Pipfile.lock` will contain version-specific/platform-specific dependencies.

Reply:
Thanks so much for explaining makes sense. I will add.